### PR TITLE
*: log each `com_stmt_fetch` separately

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

to slow log:

In the current master code, slow log will be wrong value(`last fetch time(or new start execute time) - last cmd time`)
In 2.1 code, slow log will be incompatible value(`last fetch time(or new start execute time) - first com_stmt_execute` time)
but mysql record com_stmt_exec and com_stmt_fetch for each request.

this pr will change the behavior as MySQL does, although TiDB and MySQL use very different way to handle cursor:

- MySQL execute sql and save the result in temp table(memory or myisam), then fetch the tmp table
- TiDB cache the executor-tree and execute when fetch command came and cache partial result in memory(now TiDB-server node is stateless and doesn't write any data in local disk)

so, you may see mysql stmt_execute is slow but fetch super fast, and TiDB both execute and fetch is slow in log.

### What is changed and how it works?

- add LogParitialSlow to recordset
- Fix stmt_reset should clean up  opened cursor in stmt(but pay attention both libmysql c API and jdbc, close ResultSet shouldn't send stmt_reset in current client impl, but mysql server will handle stmt_reset like this)
- log `InPreapre` and `HasMoreResult` in slow log too(help find question)
- log `Parse_time` and `Compile_time` in slow log(it seems miss those two fields) 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository

Release note

 - Write release note for bug-fix or new feature.

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/11987)

### CODE REVIEW

**Comment by lonng** [comment]
IMO, I don't like this way to fix the slow query misinformation (the extra `LogPartialSlow` method is not an elegant way). Prefer to change the query cost time to an accumulative time and accumulate it every `com_stmt_fetch`.

**Comment by lysu** [comment]
> IMO, I don't like this way to fix the slow query misinformation (the extra `LogPartialSlow` method is not an elegant way). Prefer to change the query cost time to an accumulative time and accumulate it every `com_stmt_fetch`.

mysql record slow log for each packat handle time, should keep compatible with that

https://github.com/mysql/mysql-server/blob/2f553f258daa213c710bc56d4fd7b8faeb7c4631/sql/sql_parse.cc#L1458

https://github.com/mysql/mysql-server/blob/2f553f258daa213c710bc56d4fd7b8faeb7c4631/sql/sql_parse.cc#L1738

**Comment by jackysp** [comment]
PTAL @lonng .
Please resolve the conflicts, @lysu .

File: util/sqlexec/restricted_sql_executor.go [hunk 1]
```diff
@@ -94,6 +94,10 @@ type RecordSet interface {
 	// Close closes the underlying iterator, call Next after Close will
 	// restart the iteration.
 	Close() error
+
+	// LogPartialSlow logs the slow log for partial result.
+	// it will be used in server-side cursor or multi-queries.
+	LogPartialSlow()
```
Comment (coocood) [inline util/sqlexec/restricted_sql_executor.go]: The method name is too specific.
Something like `OnFetchDone` would be better.

Comment (lysu) [reply util/sqlexec/restricted_sql_executor.go]: I tried remove this method out of that interface and give a new name now~ PTAL if free

**Comment by lysu** [comment]
@coocood @tiancaiamao  PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**